pontoon/sync/formats/json_extensions.py

```diff
@@ -0,0 +1,184 @@
+
+    @property
+    def fuzzy(self):
+        return False
+
+    @fuzzy.setter
+    def fuzzy(self, fuzzy):
+        pass  # We don't use fuzzy in JSON
+
+    @property
+    def source(self):
+        return self.data.get('placeholders', [])
+
+
+class JSONResource(ParsedResource):
+    def __init__(self, path, source_resource=None):
+        self.path = path
+        self.entities = {}
+        self.source_resource = source_resource
+
+        # Copy entities from the source_resource if it's available.
+        if source_resource:
+            for key, entity in source_resource.entities.items():
+                self.entities[key] = JSONEntity(
+                    entity.order,
+                    entity.key,
+                    copy.copy(entity.data),
+                )
+
+        try:
+            with codecs.open(path, 'r', 'utf-8') as resource:
+                self.json_file = json.load(
+                    resource,
+                    object_pairs_hook=OrderedDict
+                )
+                validate(self.json_file, SCHEMA)
+
+        except (IOError, ValueError, ValidationError) as err:
+            # If the file doesn't exist or cannot be decoded,
+            # but we have a source resource,
+            # we can keep going, we'll just not have any translations.
+            if source_resource:
+                return
+            else:
+                raise ParseError(err)
+
+        for order, (key, data) in enumerate(self.json_file.items()):
+            self.entities[key] = JSONEntity(
+                order,
+                key,
+                data,
+            )
+
+    @property
+    def translations(self):
+        return sorted(self.entities.values(), key=lambda e: e.order)
+
+    def save(self, locale):
+        """
+        Load the source resource, modify it with changes made to this
+        Resource instance, and save it over the locale-specific
+        resource.
+        """
+        if not self.source_resource:
+            raise SyncError(
+                'Cannot save JSON resource {0}: No source resource given.'
+                .format(self.path)
+            )
+
+        with codecs.open(self.source_resource.path, 'r', 'utf-8') as resource:
+            json_file = json.load(resource, object_pairs_hook=OrderedDict)
+
+            try:
+                validate(json_file, SCHEMA)
+            except ValidationError as e:
+                raise ParseError(e)
+
+        # Iterate over a copy, leaving original free to modify
+        for key, value in json_file.copy().items():
+            entity = self.entities[key]
+
+            if entity.strings:
+                json_file[key]['message'] = entity.strings[None]
+            else:
+                del json_file[key]
+
+        # Create parent directory if it doesn't exist.
+        try:
+            os.makedirs(os.path.dirname(self.path))
+        except OSError:
+            pass  # Already exists, phew!
```

Comment:
We use the same block in `silme` and `FTL`, so I'll factor it out. Note that `os.path.isdir(self.path)` will always be false, because `self.path` is the path to a resource file.